File: python/tests/transport/smbus_test_linux.py
import os
import sys

from periph.transport.smbus_linux import SMBusTransport

I2C_BUS  = int(os.environ.get('LINUX_I2C_BUS', '1'))
I2C_ADDR = int(os.environ.get('I2C_ADDR', '0x40'), 16)

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# --- offline: address validation (no bus needed) ---

class _MockBus:
    pass

try:
    SMBusTransport(_MockBus(), 0x07)
    check_true('addr 0x07 rejected', False)
except ValueError:
    check_true('addr 0x07 rejected', True)

try:
    SMBusTransport(_MockBus(), 0x78)
    check_true('addr 0x78 rejected', False)
except ValueError:
    check_true('addr 0x78 rejected', True)

# --- online: basic I/O without PEC ---

transport = SMBusTransport(I2C_BUS, I2C_ADDR)

data = transport.read(1)
check_true('read returns 1 byte', len(data) == 1)

transport.write(bytes([0x00]))
check_true('write accepted', True)

data = transport.write_read(bytes([0x00]), 1)
check_true('write_read returns 1 byte', len(data) == 1)

# --- online: write with PEC enabled ---

transport.close()
transport_pec = SMBusTransport(I2C_BUS, I2C_ADDR, pec=True)
transport_pec.write(bytes([0x00]))
check_true('write with PEC accepted', True)
transport_pec.close()

print('===DONE: {} passed, {} failed==='.format(passed, failed))
sys.exit(0 if failed == 0 else 1)

File: cpp/tests/transport/smbus_test_linux/smbus_test_linux.cpp
#include <cstdio>
#include <stdexcept>
#include "SMBusTransportLinux.h"

#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x40
#endif

static int passed = 0;
static int failed = 0;

static void check_true(const char* label, bool condition) {
    if (condition) { printf("PASS %s\n", label); passed++; }
    else           { printf("FAIL %s\n", label); failed++; }
}

int main() {
    // --- address validation ---
    try {
        SMBusTransportLinux bad(TEST_I2C_BUS, 0x07);
        check_true("addr 0x07 rejected", false);
    } catch (const std::runtime_error&) {
        check_true("addr 0x07 rejected", true);
    }

    try {
        SMBusTransportLinux bad(TEST_I2C_BUS, 0x78);
        check_true("addr 0x78 rejected", false);
    } catch (const std::runtime_error&) {
        check_true("addr 0x78 rejected", true);
    }

    // --- basic I/O without PEC ---
    SMBusTransportLinux transport(TEST_I2C_BUS, TEST_ADDR);

    uint8_t buf[1] = {0};
    transport.read(buf, 1);
    check_true("read accepted", transport.valid());

    uint8_t reg[1] = {0x00};
    transport.write(reg, 1);
    check_true("write accepted", transport.valid());

    transport.write_read(reg, 1, buf, 1);
    check_true("write_read accepted", transport.valid());

    // --- write with PEC enabled ---
    SMBusTransportLinux transport_pec(TEST_I2C_BUS, TEST_ADDR, true);
    transport_pec.write(reg, 1);
    check_true("write with PEC accepted", transport_pec.valid());

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}
